source/slang/memory_pool.cpp

```diff
@@ -0,0 +1,52 @@
+#include "memory_pool.h"
+
+namespace Slang
+{
+    const size_t kPoolSegmentSize = 4 << 20; // use 4MB segments
+
+    struct MemoryPoolSegment
+    {
+        unsigned char* data = nullptr;
+        size_t allocPtr = 0;
+        MemoryPoolSegment* nextSegment = nullptr;
+    };
+
+    MemoryPool::~MemoryPool()
+    {
+        while (curSegment)
+        {
+            auto nxtSegment = curSegment->nextSegment;
+            free(curSegment->data);
+            delete curSegment;
+            curSegment = nxtSegment;
+        }
+    }
+
+    void newSegment(MemoryPool* pool)
+    {
+        auto seg = new MemoryPoolSegment();
+        seg->nextSegment = pool->curSegment;
+        seg->data = (unsigned char*)malloc(kPoolSegmentSize);
+        pool->curSegment = seg;
+    }
+
+    void * MemoryPool::alloc(size_t size)
+    {
+        assert(size < kPoolSegmentSize);
```

Comment:
Asserting on the size is fine for the IR use case. It wouldn't be hard in the long run to have this code handle over-sized allocations by allocating a full segment for them, with a custom size.